src/Dependencies/Collections/RoslynImmutableInterlocked.cs

```diff
@@ -0,0 +1,369 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal static class RoslynImmutableInterlocked
+    {
+        /// <summary>
+        /// Mutates a value in-place with optimistic locking transaction semantics via a specified transformation
+        /// function. The transformation is retried as many times as necessary to win the optimistic locking race.
+        /// </summary>
+        /// <typeparam name="TKey">The type of key stored by the dictionary.</typeparam>
+        /// <typeparam name="TValue">The type of value stored by the dictionary.</typeparam>
+        /// <param name="location">
+        /// The variable or field to be changed, which may be accessed by multiple threads.
+        /// </param>
+        /// <param name="transformer">
+        /// A function that mutates the value. This function should be side-effect free,
+        /// as it may run multiple times when races occur with other threads.</param>
+        /// <returns>
+        /// <see langword="true"/> if the location's value is changed by applying the result of the
+        /// <paramref name="transformer"/> function; otherwise, <see langword="false"/> if the location's value remained
+        /// the same because the last invocation of <paramref name="transformer"/> returned the existing value.
+        /// </returns>
+        public static bool Update<TKey, TValue>(ref ImmutableSegmentedDictionary<TKey, TValue> location, Func<ImmutableSegmentedDictionary<TKey, TValue>, ImmutableSegmentedDictionary<TKey, TValue>> transformer)
```

Comment:
Consider representing this overload by just delegating to `Update(ref location, static (dictionary, transformer) => transformer(dictionary), transformer)`.

Reply:
@stephentoub is there a reason `ImmutableInterlocked` doesn't use the approach described by @333fred? Specifically, the question is why doesn't this method:

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs#L34

Just delegate to this method:

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs#L78

Reply:
Simply perf, in that it then requires two delegate invocations rather than one.  Given a benchmark:
```C#
    [Benchmark]
    public bool Custom() => Update1(ref _o, o => new object());
    [Benchmark]
    public bool Delegating() => Update2(ref _o, o => new object());
```
where `Update1` is what's currently checked in and `Update2` is changing it to instead delegate, on my machine I get numbers like:

|     Method |     Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |---------:|---------:|---------:|-------:|------:|------:|----------:|
|     Custom | 11.91 ns | 0.138 ns | 0.122 ns | 0.0038 |     - |     - |      24 B |
| Delegating | 16.90 ns | 0.123 ns | 0.102 ns | 0.0038 |     - |     - |      24 B |

In real situations, maybe it doesn't matter, but the function does suggest it's a pretty low-level building block where such perf concerns are relevant.

If it turns out they're not relevant and such a difference is dwarfed by other things in common usage, I'd be happy to accept a PR that simplifies it.

Reply:
➡️ For now we can keep following the pattern from dotnet/runtime